modules/binary-tuple/src/main/java/org/apache/ignite/internal/binarytuple/inlineschema/TupleWithSchemaMarshalling.java

```diff
@@ -0,0 +1,378 @@
+        System.arraycopy(valueArr, 0, result, schemaArr.length + 8, valueArr.length);
+
+        return result;
+    }
+
+    /** Get byte array from ByteBuffer without capacity overhead, only meaningful bytes are returned. */
+    private static byte[] getByteArray(ByteBuffer buff) {
+        int offset = buff.arrayOffset();
+        int limit = buff.limit();
+        byte[] result = new byte[limit - offset];
+
+        buff.get(result, offset, limit);
+        return result;
+    }
+
+    /**
+     * Unmarshal tuple (LITTLE_ENDIAN).
+     *
+     * @param raw byte[] bytes that are marshaled by {@link #marshal(Tuple)}.
+     */
+    public static @Nullable Tuple unmarshal(byte @Nullable [] raw) {
+        if (raw == null) {
+            return null;
+        }
+        if (raw.length < 8) {
+            throw new UnmarshallingException("byte[] length can not be less than 8");
+        }
+
+        // Read first int32.
+        ByteBuffer buff = ByteBuffer.wrap(raw).order(BYTE_ORDER);
+        int size = buff.getInt(0);
+        if (size < 0) {
+            throw new UnmarshallingException("Size of the tuple can not be less than zero");
+        }
+
+        // Read second int32.
+        int valueOffset = buff.getInt(4);
+        if (valueOffset < 0) {
+            throw new UnmarshallingException("valueOffset can not be less than zero");
+        }
+        if (valueOffset > raw.length) {
+            throw new UnmarshallingException(
+                    "valueOffset can not be greater than byte[] length, valueOffset: "
+                            + valueOffset + ", length: " + raw.length
+            );
+        }
+
+        ByteBuffer schemaBuff = buff
+                .position(8).limit(valueOffset)
+                .slice().order(BYTE_ORDER);
+        ByteBuffer valueBuff = buff
+                .position(valueOffset).limit(raw.length)
+                .slice().order(BYTE_ORDER);
+
+        BinaryTupleReader schemaReader = new BinaryTupleReader(size * 2, schemaBuff);
+        BinaryTupleReader valueReader = new BinaryTupleReader(size, valueBuff);
+
+        String[] columns = new String[size];
+        ColumnType[] types = new ColumnType[size];
+        Tuple tup = Tuple.create(size);
+
+        int readerInd = 0;
+        int i = 0;
+        while (i < size) {
+            String colName = schemaReader.stringValue(readerInd++);
+            int colTypeId = schemaReader.intValue(readerInd++);
+
+            columns[i] = colName;
+            types[i] = ColumnType.getById(colTypeId);
+
+            i += 1;
+        }
+
+        int k = 0;
+        while (k < size) {
+            setColumnValue(valueReader, tup, columns[k], types[k].id(), k);
+            k += 1;
+        }
+
+        return tup;
+    }
+
+    private static BinaryTupleBuilder schemaBuilder(String[] columns, ColumnType[] types) {
+        BinaryTupleBuilder builder = new BinaryTupleBuilder(columns.length * 2, columns.length * 2);
```

Comment:
```suggestion
        BinaryTupleBuilder builder = new BinaryTupleBuilder(columns.length * 2);
```

Second parameter is `totalValueSize`, in bytes. We can omit it, the builder has a built-in estimate based on element count. 

The value provided here is incorrect (too small).

Reply:
Still unresolved.